Avoid zip and zipArray issues when promises are overfulfilled

fuzzybinary opened the pull request:
Per discussion in slack, this now ignores an end user over-fulfilling a Promise and just replaces the result on an overfulfillment. However, it also avoids fulfilling zip and zipArray until all items in the dispatch group are complete by avoiding multiple group.leave() calls.

Commit:
Promises that overfulfill throw on overfulfillment

This doesn't fix all the problems with overfulfillment but makes completing a zip or zipArray without all promises in the array providing results unlikely. I've left in checks for unfulfilled promises but those should no longer happen.

Commit:
Fix issues from code review.

Fix an issue where fulfillment parameters were in the wrong place. Move overfulfillment check out of Result. Rename fulfillment barrier.

Commit:
Remove overCommit error.

Still prevent mismatched enter / leave on overfulfill but still allow overfulfillment.

Commit:
Merge branch 'develop' into bugfix/check-overfulfill

# Conflicts:
#	Sources/Promise.swift

mikerunnals reviewed (approved):
LGTM!

mikerunnals commented on Sources/Promise.swift:
```diff
@@ -234,10 +234,19 @@ public struct Promise<Value> {
+                    if !fulfilled {
+                        group.leave()
+                        fulfilled = true
+                    }
```
Maybe an `else`  here can emit the overfilfillederror, perhaps by a out-of-band callback passed in and or a logging kind of thing?

Commit:
Removed unfulfilled test

While it might still be possible for this to happen, I'm not sure how to make it happen.

samshiffman321 reviewed (approved):
lgtm!
